add code for the interrupt framework in optee-os

lackan opened the pull request:
With this commit, we add three more GIC APIs for the kernel of OPTEE-OS:
`itr_raise_sgi` : can raise software generate interrupt(SGI) from secure world to no-secure world, or secure world to secure world. It's a quick communication between different worlds and different cores. Because SGI is using the GIC N-N model, so with this API, every core can receive the interrupt if want.
`itr_raise_pi` : can trigger the peripheral interrupt with the corresponding interrupt number. When sending it to N cores, just one core can receive the effective interrupt.
`itr_set_affinity` : can target the peripheral interrupt to the core you want, it means that one can bind the interrupt to the corresponding core use this API.

The usage may as follow:
`itr_raise_sgi(11, 0x1 << 1)`
 it will raise SGI11 to core 1, and if you want not only core 1 can receive SGI11 but also core 2, then you can change the code to `itr_raise_sgi(11, 0x1 << 1 || 0x1 << 2)`
`itr_set_affinity(61, 0x1 << 1)`
`itr_raise_pi(61)`
These two APIs may use together, the operation set_affinity set the PI61 can just sent to core 1, then raise_pi, core 1 will receive the peripheral interrupt 61.

Signed-off-by: Guanchao Liang liang.guanchao@linaro.org

jenswi-linaro commented:
Hi,

Do you have any code using this interface or other means to test this?

Regards,
Jens

lackan commented:
Hi Jens,

I have tested these code on board Hi37xx, and next I may test it on QEMU.

Regards,
Guanchao.

jenswi-linaro commented:
The code doesn't compile, obviously something is missing for others to test.

lackan commented:
Yes, I have made some mistakes, I will take time to fix it. Sorry about that.

etienne-lms commented on core/drivers/gic.c:
```diff
@@ -267,6 +304,34 @@ static void gic_it_disable(struct gic_data *gd, size_t it)
 	write32(mask, gd->gicd_base + GICD_ICENABLER(idx));
 }
 
+static void gic_it_pending(struct gic_data *gd, size_t it)
```
minor: I think `gic_it_pending()` sounds like "is IT pending?".
To ease maintenance, could it be labeled `set_it_pending()` or something explicit ?

etienne-lms commented on core/drivers/gic.c:
```diff
@@ -267,6 +304,34 @@ static void gic_it_disable(struct gic_data *gd, size_t it)
+	/* Assigned to group0 */
+	assert(!(read32(gd->gicd_base + GICD_IGROUPR(idx)) & mask));
+	/* not pending yet */
+	assert(!(read32(gd->gicd_base + GICD_ICPENDR(idx)) & mask));
```
is assert() a good option here when reading the "clear pending register" ?

lackan replied:
So I just remove the assert, or I should clear the pending state before setting the interrupt pending?

etienne-lms commented on core/include/kernel/interrupt.h:
```diff
@@ -39,27 +60,50 @@ struct itr_chip {
 	void (*enable)(struct itr_chip *chip, size_t it);
+	bool (*isenabled)(struct itr_chip *chip, size_t it);
 	void (*disable)(struct itr_chip *chip, size_t it);
+	itr_result (*rpi)(struct itr_chip *chip, size_t it);
```
could you use more explicit labels for fields `rpi` `rsgi`and `target` ?
it makes code more readable.

etienne-lms commented on core/kernel/interrupt.c:
```diff
@@ -37,51 +40,196 @@
+static itr_result add_handler(size_t it, itr_handler handler,
+		uint32_t flags, void *data)
+{
+	itr_result res;
```
`res` shall be inited.

etienne-lms commented on core/kernel/interrupt.c:
```diff
@@ -37,51 +40,196 @@
+	struct itr_desc *tmp_desc;
+	uint32_t old_itr_status;
+
+	in_desc = malloc(sizeof(struct itr_desc));
```
It is best to zero-init the full structure. it prevents from forgetting to initialize each field individually.  use either `malloc();memset();` or `calloc()`.

etienne-lms commented on core/kernel/interrupt.c:
```diff
@@ -37,51 +40,196 @@
+		uint32_t flags, void *data)
+{
+	itr_result res;
+	struct itr_desc *desc = itr_descs[it];
```
should assert `it` value to prevent overflowing the `it_descs` table.

etienne-lms commented on core/kernel/interrupt.c:
```diff
@@ -37,51 +40,196 @@
+static itr_result remove_handler(size_t it, uint32_t flags, void *data)
+{
+	itr_result res;
+	struct itr_desc *desc = itr_descs[it];
```
`assert(it <= ...)`

etienne-lms commented on core/kernel/interrupt.c:
```diff
@@ -37,51 +40,196 @@
+	}
+
+	if ((desc->flags & ITRF_SHARE_MASK) == (flags & ITRF_SHARE_MASK)
+		&& (desc->flags & ITRF_SHARE_MASK) == ITRF_SHARE) {
```
minor: coding style and indentation.
```
        if ((desc->flags & ITRF_SHARE_MASK) == (flags & ITRF_SHARE_MASK) &&
           (desc->flags & ITRF_SHARE_MASK) == ITRF_SHARE) {
                if (data == NULL) {
```

etienne-lms commented on core/kernel/interrupt.c:
```diff
@@ -37,51 +40,196 @@
+itr_result itr_add(size_t it, itr_handler handler, uint32_t flags, void *data)
+{
+	itr_chip->ops->add(itr_chip, it, flags);
+	return add_handler(it, handler, flags, data);
```
i would propose the other way, to handle error case:
```
        itr_result ret;

        ret = add_handler(it, handler, flags, data);
        if (ret == GIC_OK)
                itr_chip->ops->add(itr_chip, it, flags);
        return ret;

```
if add_handler() failed, the it should be not be

lackan commented:
@etienne-lms Thanks for your comments, I will change them soon.

etienne-lms commented on core/include/kernel/interrupt.h:
```diff
@@ -30,7 +30,28 @@
+	ITRR_HANDLED,
+};
+
+typedef uint32_t itr_result;
```
do we really need a typedef for that? would be much simpler to use a `int` type.

lackan replied:
Yes, I will change.

jenswi-linaro commented on core/kernel/interrupt.c:
```diff
@@ -37,51 +41,205 @@
 
 static struct itr_chip *itr_chip;
-static SLIST_HEAD(, itr_handler) handlers = SLIST_HEAD_INITIALIZER(handlers);
+static struct itr_desc *itr_descs[256] = {0};
```
With this array you waste the larger part of 1 kB. Also there's now a fixed upper id of an interrupt.
I'd rather keep the old dynamic linked list, I don't expect that we'll have that many interrupt handlers in secure world that we'll have a problem of scaling.

lackan replied:
Get it.

jenswi-linaro commented on core/kernel/interrupt.c:
```diff
@@ -37,51 +41,205 @@
 static struct itr_chip *itr_chip;
-static SLIST_HEAD(, itr_handler) handlers = SLIST_HEAD_INITIALIZER(handlers);
+static struct itr_desc *itr_descs[256] = {0};
+static unsigned itr_lock;
```
Do we need dynamic registration/unregistration of interrupt handlers? Wouldn't it be enough to be able to enable/disable them dynamically?

jenswi-linaro commented:
This is a quite large extension of the present interrupt framework. The present interrupt framework is minimalistic so it is expected that it will be extended as needed, but I'd like to see that everything we add is needed and not just because it seems good to have in order to avoid unnecessary bloat.

lackan commented:
@jenswi-linaro Thanks, Jens. I quite agree with you! I need to pick out the really needed function in my code, and then refresh them. Looking forward  to your review at that time. What you comment help me a lot.

jenswi-linaro commented on core/drivers/gic.c:
```diff
@@ -82,11 +84,19 @@
 	.disable = gic_op_disable,
+	.raise_pi = gic_op_raise_pi,
+	.raise_sgi = gic_op_raise_sgi,
+	.affinity = gic_op_affinity,
```
`s/affinity/set_affinity/g`

jenswi-linaro commented on core/drivers/gic.c:
```diff
@@ -368,3 +404,41 @@ static void gic_op_disable(struct itr_chip *chip, size_t it)
+{
+	struct gic_data *gd = container_of(chip, struct gic_data, chip);
+
+	if (it < 16) {
```
Given that we we assert most other parameters in this API shouldn't we do the same here? The rationale is that eventual errors are static programming errors so the recovery action may be limited.

lackan replied:
OK

etienne-lms commented on core/include/kernel/interrupt.h:
```diff
@@ -32,6 +32,12 @@
 
 #define ITRF_TRIGGER_LEVEL	(1 << 0)
 
+#define cpumask_of(cpu)		(1 << cpu)
```
unused `cpumask_of()`

lackan replied:
I want to offer the user the macro cpumask_of, so it doesn't need?

jenswi-linaro replied:
One problem here where the CPU number is from, `get_core_pos()` is not guaranteed to give the number compatible with the GIC view. When reading the GIC spec it looks like the CPU numbering is implementation (platform) specific.

lackan replied:
Yes, the CPU number here just corresponding to the CPU interface. So should we complete a map here which can ensure the right CPU when get the CPU number from get_core_pos()?

jenswi-linaro commented on core/kernel/interrupt.c:
```diff
@@ -76,12 +76,32 @@ void itr_add(struct itr_handler *h)
+	itr_chip->ops->disable(itr_chip, it);
+}
+
+int itr_raise_spi(size_t it)
```
This function and `itr_raise_ppi()` are identical, wouldn't one `itr_raise_pi()` be enough?

lackan replied:
Yes, I agree.

jenswi-linaro commented on core/include/kernel/interrupt.h:
```diff
@@ -59,7 +70,21 @@ void itr_init(struct itr_chip *data);
+/* raise the PPI interrupt correspond to the interrupt ID */
+int itr_raise_ppi(size_t it);
+/*
+ * raise the SGI interrupt correspond to the interrupt ID,
```
`s/SGI interrupt/Software Generated Interrupt/`

jenswi-linaro commented on core/drivers/gic.c:
```diff
@@ -267,6 +277,32 @@ static void gic_it_disable(struct gic_data *gd, size_t it)
+static void gic_it_set_pending(struct gic_data *gd, size_t it)
+{
+	size_t idx = it / NUM_INTS_PER_REG;
+	uint32_t mask = 1 << (it % NUM_INTS_PER_REG);
```
Please use the `BIT()` or `BIT32()` macro from `<util.h>`

lackan replied:
Shouldn't here keep the same as gic_it_disable? As you recommend, I should change to "BIT32(it % NUM_INTS_PER_REG)"  instead?

jenswi-linaro replied:
`gic_it_disable()` was written before we introduced this macro.
Please change to `BIT32(it % NUM_INTS_PER_REG)` here, we can deal with `gic_it_disable()` later.
We should in general replace all these `1 << x` with `BIT(x)` or `BIT32(x)` when convenient, but in separate patches to avoid making real changes harder to read.

lackan replied:
Get it!

jenswi-linaro commented on core/drivers/gic.c:
```diff
@@ -267,6 +277,32 @@ static void gic_it_disable(struct gic_data *gd, size_t it)
+	uint32_t mask_id = it & 0xf;
+	uint32_t mask_group = group & 0x1;
+	uint32_t mask_cpu = cpu_mask & 0xff;
+	uint32_t mask = (0x0 | mask_id | (mask_group << 15) |
```
What's the purpose of `0x0`?

lackan replied:
I should remove it.

jenswi-linaro commented on core/drivers/gic.c:
```diff
@@ -368,3 +404,41 @@ static void gic_op_disable(struct itr_chip *chip, size_t it)
+	}
+
+	/* generally interrupt ID 0-7 are GROUP1, 8-15 are GROUP0 */
+	if (it < 8)
```
Please use a define instead of `8`

jenswi-linaro commented:
How can this be tested?

lackan commented:
Write a static TA to test the framework, the code include the steps as follow:

1. A interrupt handler:
static enum itr_return __maybe_unused ihandler_ok(struct itr_handler *handler)
{
        size_t core_num = get_core_pos();

        if (handler->it < SGI_NUM)
                test_sgi_value[core_num]++;
        else if (handler->it < PPI_NUM)
                test_ppi_value[core_num]++;
        else
                test_spi_value[core_num]++;

        MSG("I'm core [%zu].\n", core_num);
        MSG("I have handled the interrupt[%zu].\n", handler->it);
        return ITRR_HANDLED;
}

2. handlers for different type interrupt
struct itr_handler sgi_handler = {
        .it = TEST_SGI_ID,
        .handler = ihandler_ok,
};
struct itr_handler spi_handler = {
        .it = TEST_SPI_ID,
        .handler = ihandler_ok,
};
struct itr_handler ppi_handler = {
        .it = TEST_PPI_ID,
        .handler = ihandler_ok,
};

3. test raising the SGI/SPI/PPI
static TEE_Result test_sgi(void)
{
        itr_add(&sgi_handler);
        itr_enable(TEST_SGI_ID);

        itr_raise_sgi(TEST_SGI_ID, 0x1 << 0);
        itr_raise_sgi(TEST_SGI_ID, 0x1 << 1);
        itr_raise_sgi(TEST_SGI_ID, 0x1 << 2);
        itr_raise_sgi(TEST_SGI_ID, 0x1 << 3);

        tee_time_wait(200);
        expect_sgi_value[0]++;
        expect_sgi_value[1]++;
        expect_sgi_value[2]++;
        expect_sgi_value[3]++;
        if (memcmp(test_sgi_value, expect_sgi_value, sizeof(test_sgi_value)))
                return TEE_ERROR_GENERIC;

        itr_raise_sgi(TEST_SGI_ID, 0xff);

        tee_time_wait(200);
        expect_sgi_value[0]++;
        expect_sgi_value[1]++;
        expect_sgi_value[2]++;
        expect_sgi_value[3]++;
        if (memcmp(test_sgi_value, expect_sgi_value, sizeof(test_sgi_value)))
                return TEE_ERROR_GENERIC;

        itr_raise_sgi(TEST_SGI_ID, 0x1 << 2);
        itr_raise_sgi(TEST_SGI_ID, 0x1 << 3);

        tee_time_wait(200);
        expect_sgi_value[2]++;
        expect_sgi_value[3]++;
        if (memcmp(test_sgi_value, expect_sgi_value, sizeof(test_sgi_value)))
                return TEE_ERROR_GENERIC;

        return TEE_SUCCESS;
}
static TEE_Result test_spi(void)
{
        itr_add(&spi_handler);
        itr_set_affinity(TEST_SPI_ID, 0x1 << 3);
        itr_enable(TEST_SPI_ID);
        itr_raise_pi(TEST_SPI_ID);

        tee_time_wait(200);
        expect_spi_value[3]++;
        if (memcmp(test_spi_value, expect_spi_value, sizeof(test_spi_value)))
                return TEE_ERROR_GENERIC;

        itr_set_affinity(TEST_SPI_ID, 0x1 << 2);
        itr_raise_pi(TEST_SPI_ID);

        tee_time_wait(200);
        expect_spi_value[2]++;
        if (memcmp(test_spi_value, expect_spi_value, sizeof(test_spi_value)))
                return TEE_ERROR_GENERIC;

        itr_set_affinity(TEST_SPI_ID, 0x1 << 1);
        itr_raise_pi(TEST_SPI_ID);

        tee_time_wait(200);
        expect_spi_value[1]++;
        if (memcmp(test_spi_value, expect_spi_value, sizeof(test_spi_value)))
                return TEE_ERROR_GENERIC;

        return TEE_SUCCESS;
}
static TEE_Result test_ppi(void)
{
        itr_add(&ppi_handler);
        itr_enable(TEST_PPI_ID);
        itr_raise_pi(TEST_PPI_ID);

        tee_time_wait(200);
        expect_ppi_value[get_core_pos()]++;
        if (memcmp(test_ppi_value, expect_ppi_value, sizeof(test_ppi_value)))
                return TEE_ERROR_GENERIC;

        return TEE_SUCCESS;
}

etienne-lms commented:
Your test sounds great.
Could you add a new commit to your PR with such TA. it will be more easy to comment ?
as IMSG or DMSG rather than MSG in handler, validating other cores status when 1 core gets a spi, ...

You can rely on CFG_TEE_CORE_EMBED_INTERNAL_TESTS to build or not the static TA.

lackan commented:
@etienne-lms It's a good idea. I will finish it as you recommend.

lackan commented:
@etienne-lms Hi, some change as you recommend, but about this point "validating other cores status when 1 core gets a spi", I don't know how to do, can you please tell something in comment?

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,264 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
```
sort includes in alphabetical order.

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,264 @@
+		test_spi_value[core_num]++;
+
+	DMSG("I'm core [%zu].\n", core_num);
+	DMSG("I have handled the interrupt[%zu].\n", handler->it);
```
i think debug traces should not be called from the interrupt handlers.

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,264 @@
+{
+	itr_add(&ppi_handler);
+	itr_enable(TEST_PPI_ID);
+
```
IRQs shall be mask during this test so that its execution does not switch between cpu cores.

etienne-lms commented on core/drivers/gic.c:
```diff
@@ -248,8 +264,11 @@ static void gic_it_enable(struct gic_data *gd, size_t it)
-	/* Not enabled yet */
-	assert(!(read32(gd->gicd_base + GICD_ISENABLER(idx)) & mask));
+	if (it >= NUM_SGI)
+	/* Not enabled yet, except Software Generated Interrupt
```
rather not have comments between if `if (...)` and the related execution block.
either braces or add comment before the `if()`.

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,264 @@
+static void close_session(void *psess __unused)
+{
+}
+
```
use `CFG_TEE_CORE_NB_CORE` instead of `8`.

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,264 @@
+{
+	itr_add(&sgi_handler);
+	itr_enable(TEST_SGI_ID);
+
```
also use `CFG_TEE_CORE_NB_CORE` here so that code is portable across platforms with less than 4 cores.

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,264 @@
+#define SGI_NUM		16
+#define PPI_NUM		32
+
+#define TEST_SGI_ID	11
```
Test interrupts should be configurable (i.e from platform) so that it does conflict with platform configuration.

etienne-lms commented on core/drivers/gic.c:
```diff
@@ -368,3 +417,38 @@ static void gic_op_disable(struct itr_chip *chip, size_t it)
+	if (it >= gd->max_it)
+		panic();
+
+	/* generally interrupt ID 0-7 are GROUP1, 8-15 are GROUP0 */
```
remove this comment

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -113,34 +119,44 @@ struct itr_handler ppi_handler = {
-	itr_raise_sgi(TEST_SGI_ID, 0x1 << 3);
+	for (i = 0; i < CFG_TEE_CORE_NB_CORE; i++)
+		expect_sgi_value[i]++;
+	itr_raise_sgi(TEST_SGI_ID, 0xff);
```
mask should reflect the target cores (`SHIFT_U32(1, CFG_TEE_CORE_NB_CORE) - 1`)

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -113,34 +119,44 @@ struct itr_handler ppi_handler = {
-	expect_sgi_value[2]++;
-	expect_sgi_value[3]++;
-	itr_raise_sgi(TEST_SGI_ID, 0xff);
+	res = crypto_ops.prng.read(&num, 1);
```
this operation sends sgi to cores #0 up to core #(random).
It seems strange. 
Didn't you expect to send a SGI to a single randomly chosen target core, as implemented in `test_spi()`?

Same question for the test right below in the same routine `test_sgi()`.

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,283 @@
+
+	for (i = 0; i < CFG_TEE_CORE_NB_CORE; i++)
+		expect_sgi_value[i]++;
+	itr_raise_sgi(TEST_SGI_ID, 0xff);
```
mask should reflect the target cores `(SHIFT_U32(1, CFG_TEE_CORE_NB_CORE) - 1)`

lackan replied:
OK, I will change.

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,283 @@
+	if (memcmp(test_sgi_value, expect_sgi_value, sizeof(test_sgi_value)))
+		return TEE_ERROR_GENERIC;
+
+	res = crypto_ops.prng.read(&num, 1);
```
this operation sends sgi to cores #0 up to core #(random).
It seems strange. 
Didn't you expect to send a SGI to a single randomly chosen target core, as implemented in `test_spi()?`

Same question for the test right below in the same routine `test_sgi().`

lackan replied:
Because software generated interrupts (SGIs) using the GIC N-N model, and peripheral (hardware) interrupts using the GIC 1-N model, so for SGIs, I can send to cores #0 to core #(random) at the same time, as cpu_mask = (0x1  << 0 | 0x1 << 1 | 0x1 << 2), and use itr_raise_sgi(TEST_SGI_ID, cpu_mask) to send. Just want to distinguish from the different GIC model. What do you think?

etienne-lms commented:
@lackan, 

> @etienne-lms: Hi, some change as you recommend, but about this point "validating other cores status when 1 core gets a spi", I don't know how to do, can you please tell something in comment?

Your latest changes do cover the tests i mentioned. You already check interrupt status for all cores when trigging a interrupt to one or more cores. fine for me.

Aside from the 2 remaining comments, i am ok with this PR.

jenswi-linaro commented on core/drivers/gic.c:
```diff
@@ -248,8 +264,12 @@ static void gic_it_enable(struct gic_data *gd, size_t it)
-	/* Not enabled yet */
-	assert(!(read32(gd->gicd_base + GICD_ISENABLER(idx)) & mask));
+	if (it >= NUM_SGI) {
+		/* Not enabled yet, except Software Generated Interrupt
```
```
/*
 * Please format muliline
 * comments like this.
 */
```

lackan replied:
Get it, I will change.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

prime-zeng commented:
I have one question, should we fix the prototype of the handler in the structure from
"
struct itr_handler {
	size_t it;
	uint32_t flags;
	enum itr_return (*handler)(struct itr_handler *h);
	void *data;
	SLIST_ENTRY(itr_handler) link;
};
"
to 
"
struct itr_handler {
	size_t it;
	uint32_t flags;
	enum itr_return (*handler)(void *h);
	void *data;
	SLIST_ENTRY(itr_handler) link;
};
"
since the users of want the interrupt callback to bring back their private data pointer.

prime-zeng commented:
And we alse need to fix the itr_handle, change the h->handler(h) to h->handler(h->data);

jenswi-linaro commented:
@prime-zeng, what you're suggesting isn't exactly in the scope of this PR. Please raise an issue or open another PR to address this.

prime-zeng commented:
@jenswi-linaro  ok.

etienne-lms commented:
could we test this on qemu and qemu_v8 ?
i'm trying but get a crash. i must have forgotten something. 

jens, what about the test interrupt mapping ?
`#define TEST_SGI/PPI/SPI_ID	11/29/61`

jenswi-linaro commented:
It looks like we need safe/useful value from the platform here.
And the test case mustn't enabled unless it's supposed to work. It should at a minimum work on QEMU v7 , QEMU v8, FVP Foundation model and a suitable hardware platform with a GIC.
The reason a want so many virtual platforms is to make it easier to study the code if desired.

jenswi-linaro commented:
Doesn't build for  PLATFORM=d02

etienne-lms commented:
I tested this successfully on my plat-stm/b2260 (armv7 cortex-a9 based).
I used a unused SPI line (#250) for the test, but remained on test default SGI and PPI test lines (#11 and #29). I didn't not check how these were use by the system.

I'll will check on qemu. regarding qemu_v8, generic gic driver is not used (no GIC base addresses): to check.

etienne-lms commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,283 @@
+static size_t expect_spi_value[CFG_TEE_CORE_NB_CORE];
+static size_t expect_ppi_value[CFG_TEE_CORE_NB_CORE];
+
+static enum itr_return __maybe_unused ihandler_ok(struct itr_handler *handler)
```
This need to go in some unpaged section.
@jenswi-linaro : there seems no attributes for itr handlers to be unpaged.

jenswi-linaro replied:
Instead there's a macro `KEEP_PAGER()` that you could typically use after the function:
```
KEEP_PAGER(ihandler_ok)
```

jenswi-linaro replied:
This is needed when the pager is enabled as page faults during FIQ processing isn't supported.

lackan replied:
Get it, thanks Jens.

jbech-linaro commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (c) 2015, Linaro Limited
```
Seems like a new file, so data should probably be 2016 here instead.

jbech-linaro commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,283 @@
+	if (memcmp(test_sgi_value, expect_sgi_value, sizeof(test_sgi_value)))
+		return TEE_ERROR_GENERIC;
+
+	res = crypto_ops.prng.read(&num, 1);
```
No that it matter that much for a test, but it seems like it's the same code snippet repeating again and therefore we could shorten the code by putting it in a loop instead.

jbech-linaro commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,283 @@
+	itr_add(&spi_handler);
+	itr_enable(TEST_SPI_ID);
+
+	res = crypto_ops.prng.read(&num, 1);
```
Same here (if I'm not blindly missing anything making them different), consider using a loop instead.

lackan replied:
OK

jbech-linaro commented on core/include/kernel/interrupt.h:
```diff
@@ -59,7 +64,19 @@ void itr_init(struct itr_chip *data);
-void itr_disable(struct itr_handler *handler);
+void itr_enable(size_t it);
+void itr_disable(size_t it);
+/* raise the Peripheral Interrupt correspond to the interrupt ID */
```
s/correspond/corresponding/

lackan commented:
Hi all,

I think I've addressed all comments so far and now think this is ready for a final review and if there are no more comments we should consider merging this now.

Regards,
Guanchao.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
It looks like this PR is almost ready for merge. @jbech-linaro would you add your R-b?

jforissier commented:
@lackan please:
- Rebase on top of master
- Squash your commits (as discussed above, we want two commits: "Add interrupt framework" and "Add static TA for interrupt test")
- Add the Reviewed-by and Tested-by tags

Then I will merge this PR. Thanks!

lackan commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,259 @@
+
+#define CMD_INTERRUPT_TESTS	0
+
+#define SGI_NUM		16
```
OK, thanks Joakim.

lackan commented:
@jforissier Get it, I will finish it soon.

lackan commented on core/arch/arm/sta/interrupt_tests.c:
```diff
@@ -0,0 +1,259 @@
+#include <kernel/static_ta.h>
+#include <kernel/tee_time.h>
+#include <kernel/thread.h>
+#include <keep.h>
```
Thanks, etienne. I will change the code as you comment.

jforissier commented:
I have changed the author name to be the same as your Signed-off-by:. You should fix your Git configuration for future commits. It seems that the commit time is off, too (a few hours in the future).
I have also rebased on top of master, and merged this PR.

lackan commented:
@jforissier Thanks, I will fix them.
